algorithms/color_refinement.py

```diff
@@ -25,6 +25,48 @@ def color_refinement(G: "Graph"):
     return G
 
 
+def fast_color_refinement(G: "Graph"):
+    """
+    Performs fast color refinement on the graph.
+    It creates a queue of color groups to check for the initially colored graph.
+    For each color group in the queue, it splits the other color groups based on whether or not the vertices are
+    neighbours of one of the vertices in the color group.
+    If the result is the same for all vertices in a group, the group is left unchanged.
+    The newly created groups are added to the queue if the original group is in the queue.
+    If the original group is not in the queue, the smallest of the original and the new group is added to the queue.
+    The algorithm terminates if the queue is empty.
+    :param G: The graph to perform color refinement on
+    :return: The colored graph
+    """
+
+    queue = __initialize_queue(G)
+    while queue:
+        # Get first color and remove that from the queue
+        color = queue.pop(0)
+        # Get the vertices of the graph in the color group currently investigated
+        colors, max_colornum = get_colors(G)
+        vertices_in_color_group = colors[color]
+        # Get the neighbours of the color group grouped by color
+        neighbours_of_color_group = __get_color_groups_with_neighbours_in_color_group(vertices_in_color_group)
+        for c, color_group in neighbours_of_color_group.items():
+            # Do nothing if the size of the set in neighbours_of_color_group is zero or if the size is the same as the size of the list in colors
+            if len(color_group) == 0 or len(color_group) == len(colors[c]):
+                continue
+            # Change the color of the vertices in neighbours_of_color_group
+            max_colornum = max_colornum + 1
+            for vertex in color_group:
+                vertex.colornum = max_colornum
+            # Add the correct color to the queue
+            if c in queue:
+                queue.append(max_colornum)
+            else:
+                if 2 * len(color_group) < len(colors[c]):
```

Comment:
Is die "2*" omdat de vertices die je kleurt in colors blijven? Misschien daar even een commentje bij zetten, want over deze moest ik eerst heel goed nadenken, die is niet zo vanzelfsprekend

Reply:
Dat is om te kijken welke groep kleiner is, color_group bevat een deel van de lijst van vertices in de color group en colors[c] bevat ze allemaal. Wanneer 2* de lengte van color group nog steeds kleiner is dan de grootte van de lijst van alle vertices, dan weet ik dat die groep kleiner is dan de andere groep en wil ik die toevoegen aan de queue. Zal er iets bijzetten.

Reply:
Dit is dus precies verkeerd om zoals het hier staat, maar als het ik doe zoals in de slides duurt het algoritme langer, ik weet niet waarom.